bbb/meta-bbb/recipes-core/trebletrouble/files/src/tone.c

```diff
@@ -0,0 +1,255 @@
+#include <stdio.h>
+#include <malloc.h>
+#include <math.h>
+
+#include "libfft.h"
+#include "tone.h"
+
+#define FFT_SIZE 15
+
+/*
+ * The audio format will be PCM.
+ *  
+ * -------------Data to be determined ----------------
+ * numChannels -  1 for mono, 2 for stereo, etc...
+ * sampleRate - samples per second (44100 or 96000)
+ * bitsPerSample - 8 or 16 or 32. The higher, the better the resolution.
+ * subChunk2Size - size of the actual sound data.
+ * chunkSize - the size of the sound data + size of the header.
+ *  
+ */
+
+
+/* ------------------------------------------------------ [ Section: Endians ] */
+int isBigEndian() {
+  int test = 1;
+  char *p = (char*)&test;
+
+  return p[0] == 0;
+
+}
+
+void reverseEndianness(const long long int size, void* value) {
+
+  int i;
+  char result[32];
+  for (i = 0; i < size; i++) {
+    result[i] = ((char*)value)[size-i-1];
+  }
+  for (i = 0; i < size; i++) {
+    ((char*)value)[i] = result[i];
+  }
+}
+
+void toBigEndian(const long long int size, void* value) {
+  
+  char needsFix = !( (1 && isBigEndian()) || (0 && !isBigEndian()) );
+  if (needsFix) {
+    reverseEndianness(size,value);
+  }
+}
+
+void toLittleEndian(const long long int size, void* value) {
+  char needsFix = !( (0 && isBigEndian()) || (1 && !isBigEndian()) );
+  if (needsFix) {
+    reverseEndianness(size,value);
+  }
+}
+
+/* ------------------------------------------------------ [ Section: Wave Header ]  */
+
+
+WaveHeader makeWaveHeader(int const sampleRate, short int const numChannels, short int const bitsPerSample ) {
+
+  WaveHeader myHeader;
+
+  /* RIFF WAVE HEADER */
+  myHeader.chunkId[0] = 'R';
+  myHeader.chunkId[1] = 'I';
+  myHeader.chunkId[2] = 'F';
+  myHeader.chunkId[3] = 'F';
+  myHeader.format[0] = 'W';
+  myHeader.format[1] = 'A';
+  myHeader.format[2] = 'V';
+  myHeader.format[3] = 'E';
+
+  /* Format subchunk */
+  myHeader.subChunk1Id[0] = 'f';
+  myHeader.subChunk1Id[1] = 'm';
+  myHeader.subChunk1Id[2] = 't';
+  myHeader.subChunk1Id[3] = ' ';
+  myHeader.audioFormat = 1; /* For PCM */
+  myHeader.numChannels = numChannels; /* 1 for MONO, 2 for stereo */
+  myHeader.sampleRate = sampleRate; /* ie 44100 hertz, cd quality audio */
+  myHeader.bitsPerSample = bitsPerSample; 
+  myHeader.byteRate = myHeader.sampleRate * myHeader.numChannels * myHeader.bitsPerSample / 8;
+  myHeader.blockAlign = myHeader.numChannels * myHeader.bitsPerSample/8;
+
+  /* Data subchunk */
+  myHeader.subChunk2Id[0] = 'd';
+  myHeader.subChunk2Id[1] = 'a';
+  myHeader.subChunk2Id[2] = 't';
+  myHeader.subChunk2Id[3] = 'a';
+
+  myHeader.chunkSize = 4+8+16+8+0;
+  myHeader.subChunk1Size = 16;
+  myHeader.subChunk2Size = 0;
+
+  return myHeader;
+
+}
+
+Wave makeWave(int const sampleRate, short int const numChannels, short int const bitsPerSample) {
```

Comment:
Passing all that data on the stack is probably a bad idea.
Better to pass in a pointer like this:
```c
void makeWave(Wave* myWave, sampleRate, numChannels, bitsPerSample) {
```

Also `const int` instead of `int const` and `const short` instead of `short int const`
Note that `const` doesn't actually do anything except generate some compiler warnings if you ignore it.